### PR TITLE
abseil-dependents: Use c++17 standard

### PR DESCRIPTION
**Pull request by FlyingSamson**
#### Description

This PR is a followup to #23907 and updates the dependents of abseil to also use c++17 standard.

The dependents `Bear` and `apache-arrow` already use c++17 and `mtxclient` even c++20, so these were not updated.

I'm not sure if `protobuf-c` as a recursive dependent must be updated, too? 

Relevant ticket on trac: [https://trac.macports.org/ticket/69975](https://trac.macports.org/ticket/69975)

###### Type(s)

- [x] bugfix
- [ ] enhancement
- [ ] security fix

###### Tested on

macOS 13.6.6 22G630 arm64
Xcode 15.2 15C500b

###### Verification 
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [x] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL in commit message? 
- [x] checked your Portfile with `port lint --nitpick`?
- [ ] tried existing tests with `sudo port test`?
- [ ] tried a full install with `sudo port -vst install`?
- [ ] tested basic functionality of all binary files?
- [ ] checked that the Portfile's most important [variants](https://trac.macports.org/wiki/Variants) haven't been broken?

### CODE REVIEW

File: devel/re2/Portfile [hunk 1]
```diff
@@ -33,12 +33,13 @@ patchfiles          patch-re2pc-avoid-overlinking.diff
...
 compiler.cxx_standard \
-                    2014
+                    2017
 compiler.thread_local_storage \
                     yes
 
 configure.args-append \
-                    -DBUILD_SHARED_LIBS:BOOL=ON
+                    -DBUILD_SHARED_LIBS:BOOL=ON \
+                    -DCMAKE_CXX_STANDARD=17 \
```
Comment (barracuda156) [inline devel/re2/Portfile:42]: There is an unneeded backslash here.

File: math/s2geometry/Portfile [hunk 2]
```diff
@@ -36,12 +36,13 @@ platform powerpc {
...
 # C++ standard should match that of abseil port:
-compiler.cxx_standard   2014
+compiler.cxx_standard   2017
 
 configure.args-append \
                     -DBUILD_SHARED_LIBS=ON \
                     -DBUILD_EXAMPLES=OFF \
-                    -DBUILD_TESTS=OFF
+                    -DBUILD_TESTS=OFF \
+                    -DCMAKE_CXX_STANDARD=17 \
```
Comment (barracuda156) [inline math/s2geometry/Portfile:45]: Same

**Review by barracuda156 — CHANGES_REQUESTED** [review]

File: devel/re2/Portfile (hunk 1 above)
Comment (FlyingSamson) [reply devel/re2/Portfile]: Thank you. I removed them.

File: math/s2geometry/Portfile (hunk 2 above)
Comment (FlyingSamson) [reply math/s2geometry/Portfile]: Fixed

**Comment by barracuda156** [comment]
Those failing `py-grpcio` ports are known to be forever broken.

File: math/s2geometry/Portfile [hunk 3]
```diff
@@ -36,12 +36,13 @@ platform powerpc {
...
 # C++ standard should match that of abseil port:
-compiler.cxx_standard   2014
+compiler.cxx_standard   2017
 
 configure.args-append \
                     -DBUILD_SHARED_LIBS=ON \
                     -DBUILD_EXAMPLES=OFF \
-                    -DBUILD_TESTS=OFF
+                    -DBUILD_TESTS=OFF \
+                    -DCMAKE_CXX_STANDARD=17
```
Comment (reneeotten) [inline math/s2geometry/Portfile:45]: shouldn't this use the option `cmake.set_cxx_standard` from the PortGroup instead of adding it manually?

Comment (FlyingSamson) [reply math/s2geometry/Portfile:45]: I wasn't aware of that option and just copied what I saw in the change of abseil itself. 
I think you are right. Should I then place this directly below the calls to `compiler.cxx_standard`?

**Comment by FlyingSamson** [comment]
Should we also address the issue described in [ticket 69975](https://trac.macports.org/ticket/69975) in this PR?

**Comment by ryandesign** [comment]
> * [x]  referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL in commit message?

You have not referenced the ticket in the commit message. We can fix it by using the Squash and merge button when we merge this PR.

> Should we also address the issue described in [ticket 69975](https://trac.macports.org/ticket/69975) in this PR?

Doesn't this PR already exactly address that ticket? and the commit message should therefore say that it closes that ticket?

> Those failing `py-grpcio` ports are known to be forever broken.

Hopefully not forever, but the ticket about the problem is https://trac.macports.org/ticket/69262.

**Comment by FlyingSamson** [comment]
> Doesn't this PR already exactly address that ticket? and the commit message should therefore say that it closes that ticket?

Sorry, I somehow seem to have thought that there was still more to be done to fix that ticket. But I cannot see why this would be the case anymore.

> We can fix it by using the Squash and merge button when we merge this PR.  

If that makes things easier I can also amend my commit message with the ticket URL and push it again.

**Comment by FlyingSamson** [comment]
I now included the two trac tickets that are addressed by this pull request in the commit message.

**Comment by reneeotten** [comment]
I doubt the `cmake.set_cxx_standard yes` is required explicitly in all the ports.... You claimed something failed and I gave a suggestion to look into, not to just blindly add everywhere. Please wrap this up so it can be merged or close the PR.

**Comment by FlyingSamson** [comment]
I’m sorry, but I an not sure what you are referring to when you say
> You claimed something failed and I gave a suggestion to look into, not to just blindly add everywhere

I initially did not have the `cmake.set_cxx_standard` set but instead included `-DCMAKE_CXX_STANDARD=17` in the configure args of these ports. You then correctly pointed out to me, that instead I should rather use the port group option. So I did not blindly apply this flag but switched from passing the standard as an argument to cmake to using the port group option. 

As for the necessity of requiring cmake to configure those ports using c++17 standard: I went through the dependent ports of `abseil`. Some already required c++17 or newer and were hence not altered (see PR description above). The source code of all ports changed by this PR (`re2`, `protobuf`, `grpc`, and `s2geometry`) all use some `abseil` feature which is now available in the c++17 standard (e.g., `std::string_view` and `std::optional`) and `abseil` will hence forward to the implementations in the standard (because the `abseil` port is now build with c++17). This should lead to compile errors for these ports such as the ones for `re2` in the now linked trac tickets when compiled with a standard prior to c++17 (I could not reproduce this error on my system, presumably because the issue opener used an older system/compiler version, while the compiler on my system might default to c++17, evading that problem).


> Please wrap this up so it can be merged or close the PR.  

If you think it preferable to keep the remaining ports (`protobuf`, `grpc`, and `s2geometry`) on their previous c++ standard versions and wait if this indeed causes issues in the long run, I can also revert those changes for now.

**Comment by FlyingSamson** [comment]
I finally managed to get a virtual machine up and running with macOS 10.12 and Xcode 9.2, as described in the trac ticket.

Testing in that image I found, that indeed the use of `cmake.set_cxx_standard yes` was not required because the flag for `c++17` already is a public dependency of the `abseil` cmake config file. So I removed those.

Unfortunately, I also found that the link stage of `grpc` and `protobuf3-cpp-upstream` fail due to missing symbols in the standard library (e.g., the destructor of `bad_optional_access`). This has been deemed as being a bug in the apple clang standard library implementation [here](https://github.com/tcbrindle/cpp17_headers/pull/12). I'm not sure if and how this can be handled in those ports. Should these include `legacysupport.use_mp_libcxx yes`?

**Pull request merged**